### ISSUE TITLE
Incorrect handling of offset commits after a rebalance

### ISSUE THREAD

**Issue by moreda** [splunk/kafka-connect-splunk#260]
Hi.

After a rebalance –with topic/partition reassignment– some tasks could try to commit offsets in partitions not assigned to them. This happens when some acknowledgements arrive to a task and the partition is not assigned to it anymore.

In such cases, the task won't be able to commit the offset of that topic/partition, and you'll see messages like:

```
WARN WorkerSinkTask{id=kafka-connect-splunk-1} Ignoring invalid task provided offset syslog-3/OffsetAndMetadata{offset=117798, leaderEpoch=null, metadata=''} -- partition not assigned, assignment=[syslog-1] (org.apache.kafka.connect.runtime.WorkerSinkTask:422)
```

It doesn't seem critical, because the task that receives the new partition assignment will try to send again to Splunk HEC and commit eventually. The problem is that the task with the previous assignment won't free memory and it's going to try again and again to commit an offset in an unassigned partition.

A simple way to avoid this situation is to use the info provided by `SinkTask.open()` and `SinkTask.close()` to purge and filter events when a partition reassignment occurs. This purge has to be made in buffered events, failed events, offsets, and topic/partition records. This should be considered safe because those events will be picked up by the task that opens the partitions closed in the original task.

Besides this main issue, I've found another issue related to the fact that there's no guarantee that a batch contains events tied to a single topic/partition. In `KafkaRecordTracker.removeAckedEventBatch()` just the topic/partition of the first event is considered (`final Event event = events.get(0);`). The consequence of this is that offsets are not committed as soon as they can, but depending on the "randomness" of the first event topic/partition.  

To solve all this I'm proposing the PR #259 for review that (I think) solves issue. I extended the tests, I tried it in a small environment and it seems to work as expected. 😄 
Cheers,

  Rob

**Comment by moreda** [comment]
Sorry, in the PR I included an extra commit with small fix in the docs that caused me some headaches when testing 😊

**Comment by moreda** [comment]
I rebased the PR #259 on the current develop branch and didn't include the doc fix commit (less noise).

I'm aware that this is a "non trivial" PR, but it's important for things to work properly in situations where Kafka Connect instances could go down, number of tasks increase/decrease or simply software is being updated. Not taking this in account could lead to lots of memory not garbage collected, besides umpteen warning messages as mentioned above (that was the indicator for me that something was wrong).

IMHO, labelling this issue just as an "enhancement" is confusing because in a normal architecture (several tasks and horizontal elasticity) this leads to actual problems.

If you need a different approach or more details on what those problems are, I'll be happy to help. Of course, I could be missing the whole point and be wrong at many levels too 😄.

**Comment by waewoo** [comment]
@chaitanyaphalak 
hello, 

We are also affected by this bug.
When can you expect this PR to be merged?

Thanks

**Comment by rockb1017** [comment]
Thank you for your work. I will test it and verify. 
so this rebalance is when one or more of kafka connects shut down, right? 

so a proper way to test this for example would be: 
1. set up a topic with 3 partitions
2. launch 3 kafka connects with same consumer group 
3. generate messages at high load to the topic
4. shutdown one of kafka connects. 
5. then verify:
* there is no infinite retry of committing ack
* there is no data loss

is this correct?

**Comment by moreda** [comment]
Hi, @rockb1017.
You're absolutely correct :-).
If you need more detail on some parts please don't hesitate to ask.

**Comment by rockb1017** [comment]
looks good to me and i validated that it is working with no errors. Merging it. thank you very much!

**Comment by moreda** [comment]
You're welcome and happy to help. Now waiting impatiently for a the new GA release (any ETA?) ;-)
Thanks a lot for your work!

**Comment by mcascallares** [comment]
@moreda it's released in v2.0.2 https://docs.splunk.com/Documentation/KafkaConnect/2.0.2/User/ReleaseNotes
Thank you!
